Add delete-by-query plugin

tlrx opened the pull request:
This pull request adds a new plugin called "delete-by-query" which implements the now deprecated delete-by-query feature using scan/scroll/bulk requests.

Notes:
* `size` parameter controls the scroll shard_size and the number of actions in bulk requests (defaults to 1000)
* `timeout` parameter can be used to stop scrolling documents after a given time
* response now looks like this (here a node is killed during the DBQ execution):
```json
{  
   "took":60866,
   "timed_out":false,
   "_indices":{  
      "_all":{  
         "found":531046,
         "deleted":79901,
         "missing":0,
         "failed":301702
      },
      "disposants-2014":{  
         "found":375702,
         "deleted":74000,
         "missing":0,
         "failed":301702
      },
      "beer":{  
         "found":5901,
         "deleted":5901,
         "missing":0,
         "failed":0
      }
   },
   "failures":[  
      {  
         "shard":-1,
         "index":null,
         "reason":{  
            "type":"node_not_connected_exception",
            "reason":"[Puck][inet[/192.168.1.16:9300]] Node not connected"
         }
      }
   ]
}
```

Since the process involves the execution of a scan request (which can fail), then successive async scroll requests (which can also fail) we may imagine a better failure reporting. When a scroll request succeed, the scrolled documents are added to a Bulk request executed in an async manner. If the bulk fails, all documents are reported as `failed` documents in the counter.

Rest API documentation and test will be added later.

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,454 @@
+                    scanRequest.routing(request.routing());
+                }
+
+                SearchSourceBuilder source = new SearchSourceBuilder().query(request.source()).fields("_routing", "_parent").fetchSource(false).version(true);
```
Can we check if we need to get `_parent` as well? I can't recall if we use it as routing without storing it in the _routing field, I think we do always also put it in _routing, need to check

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,454 @@
+                }
+                scanRequest.source(source);
+
+                logger.debug("executing scan request");
```
this probably need to be trace, action package has DEBUG enabled by default

kimchy replied:
applies to other logging statements here

tlrx replied:
Done

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,454 @@
+        }
+
+        void executeScroll(final String scrollId) {
+            threadPool.generic().execute(new AbstractRunnable() {
```
I don't think we need to have it execute on another thread pool, its perfectly fine just to go and execute it right away, it doesn't block on anything

tlrx replied:
Ok

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,454 @@
+            }
+
+            // Delete the scrolled documents using the Bulk API
+            threadPool.generic().execute(new AbstractRunnable() {
```
I don't think we need to execute it on a thread pool, nothing blocking, we can do it without one and delegate to the async bulk execution

tlrx replied:
Right

kimchy commented:
I left some minor comments around logging and usage of thread pool (not needed I think).

I could't follow why we need a semaphore and such, I think I a missing something. My thought was that we do search -> bulk -> search -> .... until there are no more results, so always async callback execution type chain until we are done.

tlrx commented:
@kimchy thanks for your review! Your comments make sense, no need to use semaphore stuff... I rebased and updated the code, it is way simpler now.

I'm adding some rest tests too.

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,350 @@
+                    final String nextScrollId = scrollResponse.getScrollId();
+                    addShardFailures(scrollResponse.getShardFailures());
+
+                    if (logger.isDebugEnabled()) {
```
this should be trace?

tlrx replied:
raaaah yes

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,350 @@
+        }
+
+        void onBulkResponse(int bulkId, String scrollId, BulkResponse bulkResponse) {
+            if (logger.isDebugEnabled()) {
```
trace? I don't think we need the check, btw (isDebug)

kimchy replied:
this applies to the rest of the logging statements here, so won't comment on those as well

s1monw commented on plugins/delete-by-query/pom.xml:
```diff
@@ -0,0 +1,46 @@
+
+    <properties>
+        <!-- You can add any specific project property here -->
+        <tests.jvms>1</tests.jvms>
```
why do we need 1 JVM here only?

tlrx replied:
Wrong copy paste, we don't need to limit to 1 JVM.

kimchy commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/IndexDeleteByQueryResponse.java:
```diff
@@ -0,0 +1,142 @@
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        super.readFrom(in);
+        index = in.readString();
```
we need to serialize the counters here as well

tlrx commented:
@s1monw thanks for your review. I updated the code following your comment and added a REST test. Can you please have another look if possible? Thanks :)

Documentation will be added in another PR.

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+            out.writeBoolean(false);
+        } else {
+            out.writeBoolean(true);
+            out.writeVLong(timeout);
```
if you write a VLong make sure it's not negative!

tlrx replied:
Ok, `timeout` has been changed to `TimeValue`

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+
+    private Scroll scroll = new Scroll(TimeValue.timeValueMinutes(10));
+
+    private Long timeout;
```
IMO timeout should be a `TimeValue`

tlrx replied:
Ok, `timeout` has been changed to `TimeValue`

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+
+    private String routing;
+
+    private Integer size = 1_000;
```
can this be `int` rather than `Integer`?

tlrx replied:
Sure.

I also think that a default size of 1000 docs per shard is not a good idea... Let's use the default value of `10`.

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    public DeleteByQueryRequest indicesOptions(IndicesOptions indicesOptions) {
+        if (indicesOptions == null) {
```
maybe validate size and timeout too here?

tlrx replied:
Ok

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+        out.writeStringArray(types);
+        out.writeBytesReference(source);
+        out.writeOptionalString(routing);
+        out.writeVInt(size);
```
watach out size must be positive here though

tlrx replied:
Ok. `size` is initialized to 0 and can't be set to a negative value.

For my education, `writeVInt` does not support negative values at all or it's just a question of better serialization?

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+        out.writeBytesReference(source);
+        out.writeOptionalString(routing);
+        out.writeVInt(size);
+        if (scroll == null) {
```
you can use `writeOptionalStreamable` here?

tlrx replied:
Ok

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,264 @@
+        } catch (Exception e) {
+            // ignore
+        }
+        return "[" + Arrays.toString(indices) + "][" + Arrays.toString(types) + "], source[" + sSource + "]";
```
maybe prepend that this is a delete by query?

tlrx replied:
Ok

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/IndexDeleteByQueryResponse.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    IndexDeleteByQueryResponse(String index) {
+        super();
```
super crot is not needed

tlrx replied:
Right

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/IndexDeleteByQueryResponse.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    public IndexDeleteByQueryResponse(String index, long found, long deleted, long missing, long failed) {
+        this.index = index;
```
call `this(index)` here?

tlrx replied:
Yes

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/IndexDeleteByQueryResponse.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    public void incrementFound(long delta) {
+        this.found = found + delta;
```
can assert here everywhere that they never become negative

tlrx replied:
Sure. I also added some unit tests for that.

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,307 @@
+        AsyncDeleteByQueryAction(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+            this.startTime = System.currentTimeMillis();
```
maybe we wanna use `ThreadPool.estimatedTimeInMillis()` here?

tlrx replied:
I didn't know about this class but for what I saw it captures the system current time at regular intervals, trading off a less time precision for better performance, right?

If so, I think we can use it here since millisecond precision for this potentially long process is not really necessary.

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,307 @@
+                        return;
+                    }
+
+                    if (isTimedOut()) {
```
can we call it `hasTimedOut`?

tlrx replied:
Yes

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,307 @@
+        }
+
+        private boolean isTimedOut() {
+            return request.timeout() != null && (System.currentTimeMillis() >= (startTime + request.timeout().millis()));
```
also use the Threadpool estimations here mabye?

tlrx replied:
Yes

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/plugin/deletebyquery/DeleteByQueryPlugin.java:
```diff
@@ -0,0 +1,49 @@
+    public Collection<Module> modules(Settings settings) {
+        Collection<Module> modules = new ArrayList<>();
+        modules.add(new DeleteByQueryModule());
+        return modules;
```
just call `return Arrays.asList(new DeleteByQueryModule())`

tlrx replied:
Ok

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,307 @@
+                        listener.onFailure(e);
+                    }
+                });
+            } catch (Throwable t) {
```
I think you should have such a catch clause in every method in here since might easily hang the entire requires if there is a simple assertion tripped elsewhere

tlrx replied:
I agree, good catch, I should have thought of that. I reuse the `finishHim` method when possible in order to clean the `scrolIId` and release search context and report failures.

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,307 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class TransportDeleteByQueryAction extends HandledTransportAction<DeleteByQueryRequest, DeleteByQueryResponse> {
```
If possible I'd love to have a single node test that really calls all these funtions directoy to ensure we tests all the corner cases, makse sense?

s1monw replied:
can you document this class and state what the capabilities of this plugin are? I think we should make clear to the user that this is best effort, non-transactional, memory intensive etc. etc. We should maybe also think about having a shortcut if there is a match all query passed to this. I think folks to this all the time so I wonder if we should have such a destructive operation as feature on this as well behind the scenes. Doesn't need to be added now but in general we can think about adding this to IndexShard / Engine since in lucene that is quite fast.

tlrx replied:
I created the `TransportDeleteByQueryActionTests` to test the methods, let me know what you think about it.

> We should maybe also think about having a shortcut if there is a match all query passed to this. I think folks to this all the time so I wonder if we should have such a destructive operation as feature on this as well behind the scenes. Doesn't need to be added now but in general we can think about adding this to IndexShard / Engine since in lucene that is quite fast.

I created an issue for that: #11700

s1monw commented on plugins/delete-by-query/src/main/java/org/elasticsearch/plugin/deletebyquery/DeleteByQueryModule.java:
```diff
@@ -0,0 +1,48 @@
+import org.elasticsearch.action.deletebyquery.TransportDeleteByQueryAction;
+import org.elasticsearch.rest.action.deletebyquery.RestDeleteByQueryAction;
+import org.elasticsearch.rest.RestModule;
+
```
can we have a doc-string here explaining what this thing does and why. Ie. I'd also like to have some docs in here that explain why it's not in core anymore etc. if you need help I am happy to explain

s1monw commented:
looks pretty good though. I left a bunch of comments

tlrx commented:
@s1monw thanks a lot for your review, very valuable. I updated the code following your comments, please let me know if there are still things to improve.

I'd love to have your help on writing documentation for this plugin, since I'm not sure to be able to explain all fallacies of the previous implementation.

s1monw commented:
> I'd love to have your help on writing documentation for this plugin, since I'm not sure to be able to explain all fallacies of the previous implementation.

lets get this in as is and open an issue for the documentation I will take a look at comment on it what aspects I would take into account?

s1monw commented:
oh yeah so here is my LGTM ;)

tlrx commented:
@s1monw thanks!

I created https://github.com/elastic/elasticsearch/issues/11723 for the java doc aspect.
